ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py

```diff
@@ -800,6 +924,139 @@ def process(self, device, results, unused):
                     status=floatingip.get('status', 'UNKNOWN'),
                     )))
 
+        # volumes
+        volumes = []
+        for volume in results['volumes']:
+            if not volume.get('id', None):
+                continue
+
+            attachment = volume.get('attachments', [])
+            instanceId = ''
+            if len(attachment) > 0:
+                instanceId = attachment[0].get('server_id', '')
+            data=dict(
+                id=prepId('volume-{0}'.format(volume['id'])),
+                title=volume.get('name', ''),
+                volumeId=volume['id'],  # 847424
+                avzone=volume.get('availability_zone', ''),
+                created_at=volume.get('created_at', '').replace('T', ' '),
+                sourceVolumeId=volume.get('source_volid', ''),
+                host=volume.get('os-vol-host-attr:host', ''),
+                size=volume.get('size', 0),
+                bootable=volume.get('bootable', 'FALSE').upper(),
+                status=volume.get('status', 'UNKNOWN').upper(),
+                set_tenant=prepId('tenant-{0}'.format(volume.get('os-vol-tenant-attr:tenant_id',''))),
+            )
+            # set instance only when volume['attachments'] is not empty
+            if len(instanceId) > 0:
+                data['set_instance'] = prepId('server-{0}'.format(instanceId))
+            volumes.append(ObjectMap(
+                modname='ZenPacks.zenoss.OpenStackInfrastructure.Volume',
+                data=data))
+
+
+        # Snapshots
+        snapshots = []
+        for snapshot in results['snapshots']:
+            if not snapshot.get('id', None):
+                continue
+
+            snapshots.append(ObjectMap(
+                modname='ZenPacks.zenoss.OpenStackInfrastructure.Snapshot',
+                data=dict(
+                    id=prepId('snapshot-{0}'.format(snapshot['id'])),
+                    title=snapshot.get('name', ''),
+                    created_at=snapshot.get('created_at', '').replace('T', ' '),
+                    size=snapshot.get('size', 0),
+                    description=snapshot.get('description', ''),
+                    status=snapshot.get('status', 'UNKNOWN').upper(),
+                    set_volume=prepId('volume-{0}'.format(snapshot.get('volume_id',''))),
+                    set_tenant=prepId('tenant-{0}'.format(snapshot.get('os-extended-snapshot-attributes:project_id',''))),
```

Comment:
Same concern for this one.  I would think we can safely assume that volume_id will be present, but any of these extended attributes might be missing, depending on the backend.   Better to avoid the relationship than to set it to a bogus value, which will fail.